File: void_front/game/core/player/ships/enemies/sniper.py
from __future__ import annotations

from game.core.assets.main import assets
from game.core.player.ships.enemies.enemy import EnemyShip
from game.core.player.ships import player
from game.core.player.ships.shapes import shapes
from game.core.weapons.heavy_rail_gun import HeavyRailGun
from game.core.world.utils import toroidalDistance, updateWrapped
import numpy as np


class Sniper(EnemyShip):
    def __init__(self, x=10, y=20, angle=0):
        super().__init__(
            x=x,
            y=y,
            width=50,
            height=50,
            angle=angle,
            acceleration=600,
            color="red",
            vertices=shapes[3],
            name="Sniper",
            maxWeaponHeat=20000,
            life=100,
            weapon=HeavyRailGun,
            img=getattr(getattr(assets, "images", None), "snipership", None),
            flameImg=getattr(getattr(assets, "images", None), "flame1", None),
        )
        self.seekAcceleration = self.acceleration
        self.fleeAcceleration = self.acceleration * 0.9

    def update(self, t, dt, thrust=0, turn=0):
        updateWrapped(
            fn=lambda *_: self._update(t, dt, thrust=thrust, turn=turn),
            x=self.x,
            y=self.y,
            margin={"width": 5000, "height": 5000},
        )

    def _update(self, t, dt, thrust=0, turn=0):
        super().update(t, dt, thrust=thrust, turn=turn)
        dist = toroidalDistance(self.x, self.y, player.ship.x, player.ship.y)
        self.AI(
            idleDistance=dist > 11**11,
            fleeCondition=dist < 150000,
            seekCondition=(self.state == "flee" and dist > 8**10)
            or self.state != "flee",
            fireCondition={
                "func": lambda delta: abs(delta) < np.pi / 32,
                "others": dist < 10000000,
            },
        )

File: core/player/ships/enemies/sniper.js
import { assets } from "../../../assets/main.js";
import { clamp } from "../../../utils/math.js";
import HeavyRailGun from "../../../weapons/heavyRailGun.js";
import { canvas } from "../../../world/canvas.js";
import { toroidalAngle, toroidalDistance, updateWrapped } from "../../../world/utils.js";
import { ship } from "../player.js";
import { shapes } from "../shapes.js";
import EnemyShip from "./enemy.js";

const rTD = (r) => 180 / Math.PI * r;
export default class Sniper extends EnemyShip {
    constructor({ x = 10, y = 20, angle = 0 }) {
        super({ x, y, width: 50, height: 50, angle, acceleration: 1000, color: "red", vertices: shapes[3], name: "Sniper", maxWeaponHeat: 20000, life: 100, weapon: HeavyRailGun, img: assets?.images?.snipership, flameImg: assets?.images?.flame1 });
        this.seekAcceleration = this.acceleration;
        this.fleeAcceleration = this.acceleration * 0.9;
    }

    update(t, dt) {
        updateWrapped({
            fn: () => {
                const dist = toroidalDistance(this.x, this.y, ship.x, ship.y);
                super.update(t, dt);
                this.AI({
                    idleDistance: dist > 11 ** 11, fleeCondition: dist < 150000, seekCondition: (this.state === "flee" && dist > 8 ** 10) || this.state !== "flee", fireCondition: {
                        func: (delta) => Math.abs(delta) < Math.PI / 32,
                        others: dist < 10000000,
                    }
                })
            }, x: this.x, y: this.y, margin: {
                width: 5000,
                height: 5000
            }
        })
    }
}